lib/Request.cpp

```diff
@@ -0,0 +1,328 @@
+        return matches;
+    }
+
+    bool Request::match(string pattern)
+    {
+        key = method + ":" + url;
+        return regex_match(key, matches, regex(pattern));
+    }
+#endif
+
+
+    bool Request::hasCookie(const std::string &key) const
+    {
+        int i;
+        char dummy[10];
+
+        for (auto header: mHeaders)
+        {
+            if (header.first.find("Cookie") != std::string::npos)
+            {
+                if (mg_get_cookie(header.second.c_str(), key.c_str(), dummy, sizeof(dummy)) != -1)
+                {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    std::string Request::getCookie(const std::string &key, const std::string &fallback) const
+    {
+        std::string output;
+        int i;
+        int size = 1024;
+        int ret;
+        char *buffer = new char[size];
+        char dummy[10];
+        const char *place = NULL;
+
+        for (auto header: mHeaders)
+        {
+            if (header.first.find("Cookie") != std::string::npos)
+            {
+                if (mg_get_cookie(header.second.c_str(), key.c_str(), dummy, sizeof(dummy)) != -1)
+                {
+                    place = header.second.c_str();
+                    break;
+                }
+            }
+        }
+
+        if (place == NULL) {
+            return fallback;
+        }
+
+        do {
+            ret = mg_get_cookie(place, key.c_str(), buffer, size);
+
+            if (ret == -2) {
+                size *= 2;
+                delete[] buffer;
+                buffer = new char[size];
+            }
+        } while (ret == -2);
+
+        output = std::string(buffer);
+        delete[] buffer;
+
+        return output;
+    }
+
+    bool Request::hasHeader(const std::string &key) const
+    {
+        return mHeaders.find(key) != mHeaders.end();
+    }
+
+    std::string Request::getHeaderValue(const std::string& key) const
+    {
+      std::string result;
+
+      if (hasHeader(key))
+      {
+          result = mHeaders.at(key);
+      }
+
+      return result;
+    }
+
+    std::map<std::string, std::string> Request::headers() const
+    {
+        return mHeaders;
+    }
+
+    std::string Request::url() const
+    {
+        return mUrl;
+    }
+
+    std::string Request::method() const
+    {
+        return mMethod;
+    }
+
+    std::string Request::body() const
+    {
+        return mBody;
+    }
+
+    bool Request::isValid() const
+    {
+        return mIsValid;
+    }
+
+    void Request::setIsValid(bool value)
+    {
+        mIsValid = value;
+    }
+
+    bool Request::isMultipartRequest() const
+    {
+        return mIsMultipartRequest;
+    }
+
+    std::vector<Request::MultipartEntity> Request::multipartEntities() const
+    {
+        return mMultipartEntities;
+    }
+
+    void Request::setMultipartEntities(const std::vector<Request::MultipartEntity> &entities)
+    {
+        mMultipartEntities = entities;
```

Comment:
What if two threads are calling this function? Is it possible?

Reply:
Technically it *is* possible. So maybe we should ask the users to not do that.

This is mostly done after we finished getting a multipart request, before we call Controller::process (which is where a user can spawn new threads if they need).
A one time call really.